### ISSUE TITLE
asyncio: selector_events: add_urgent() for urgent data to read, 3rd argument of select.select()

### ISSUE THREAD

**Issue by b5372eda-6675-4503-a2f8-d774d4d4f105** [python/cpython#75030]
BPO | [30847](https://bugs.python.org/issue30847)
--- | :---
Nosy | @gvanrossum, @vstinner, @1st1, @pklanke

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2017-07-04.13:51:23.279>
labels = ['type-feature', '3.7', 'expert-asyncio']
title = 'asyncio: selector_events: add_urgent() for urgent data to read, 3rd argument of select.select()'
updated_at = <Date 2017-07-06.11:22:47.102>
user = 'https://github.com/pklanke'
```

bugs.python.org fields:
```python
activity = <Date 2017-07-06.11:22:47.102>
actor = 'vstinner'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['asyncio']
creation = <Date 2017-07-04.13:51:23.279>
creator = 'pklanke'
dependencies = []
files = []
hgrepos = []
issue_num = 30847
keywords = []
message_count = 9.0
messages = ['297652', '297653', '297654', '297655', '297722', '297765', '297809', '297821', '297823']
nosy_count = 4.0
nosy_names = ['gvanrossum', 'vstinner', 'yselivanov', 'pklanke']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue30847'
versions = ['Python 3.7']
```

</p></details>

**Comment by b5372eda-6675-4503-a2f8-d774d4d4f105** [comment]
depends on bpo-30844

**Comment by vstinner** [comment]
Please elaborate the description :-)

**Comment by vstinner** [comment]
Does the Windows Proactor event loop has a similar concept of "excepter"?

**Comment by b5372eda-6675-4503-a2f8-d774d4d4f105** [comment]
To be able to use GPIO Sysfs Interface with asyncio on our embedded platforms, we require exceptional event support in asyncio.

depends on bpo-30844

**Comment by vstinner** [comment]
In the selectors issue, we discussed how an application should prioritize "urgent" events:
http://bugs.python.org/issue30844#msg297707

While I now agree that it's not the role of selectors to decide, I would like to discuss the plan for asyncio.

Let's say that we got read event on sockets A and B (in an ordered list from selectors: A, then B), but B gets urgent data: should we handle B urgent data before not-urgent A data?

Would it be possible to let the developer decide how to prioritize events?

How does Twisted or Node.JS handle urgent data?

**Comment by gvanrossum** [comment]
How do you define "urgent data"? Is this just the third category of select(), Read, Write, Exceptional? I don't know if that should be considered urgent, it's just "out of band" IIRC.

**Comment by b5372eda-6675-4503-a2f8-d774d4d4f105** [comment]
This is in fact the third catagory of select(), "exceptional conditions", but because some find the term "exceptional" confusing when used in a Python module, we decided to use the term "urgent data", borrowed from the poll(2) man page. (see bpo-30844)

An example of an exceptional conditions event is "out of band", another one is edge detection from GPIO Sysfs Interface.

The patch for fulfilling this issue can and will be supplied if and when bpo-30844 is merged. 30844 originally also contained the patch to asyncio, but I was asked to split it into two separate patches.

**Comment by b5372eda-6675-4503-a2f8-d774d4d4f105** [comment]
> "Let's say that we got read event on sockets A and B (in an ordered list from selectors: A, then B), but B gets urgent data: should we handle B urgent data before not-urgent A data?"

IMO No. The same strategy applies. urgent data events on B have priority over other events on B, but not necessarily over events on A. As long as the urgent data events for a certain file object are handled before other events for that file object, it should be fine.

> "Would it be possible to let the developer decide how to prioritize events?"
At this point asyncio has no support for prioritizing event handlers. If prioritizing should be necessary, it should IMO be implemented in a way that all event handlers can be prioritized and not just these events. I think we can agree that this falls outside the scope of this patch.

**Comment by vstinner** [comment]
Ok, it makes sense. Thanks :-)
